src/ImageSharp/ColorProfiles/Icc/Calculators/ColorTrcCalculator.cs

```diff
@@ -39,13 +39,16 @@ public ColorTrcCalculator(
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
     public Vector4 Calculate(Vector4 value)
     {
+        // uses the descaled XYZ as DemoMaxICC IccCmm.cpp : CIccXformMatrixTRC::Apply()
+        Vector4 xyz = new(CieXyz.FromScaledVector4(value).ToVector3(), 1);
```

Comment:
Amusingly the matrix is applied to the actual XYZ not the scaled XYZ. Is there a better way to descale this?

Reply:
Hmmmm… this means we are scaling then descaling each pixel. Do all other calculators use scaled values? If so can we avoid the initial scaling or update this to match?

Reply:
This is the only calculator I've encountered so far that requires the non-scaled values. Won't be sure until we find profiles with the other types of curves but it's likely to be just this one, for the matrix multiplication to apply correctly.

We might be able to do something similar to the `Is16BitLutEntry` check and avoid scaling XYZ if we know we're using `ColorTrcCalculator`? It would add more complexity to the target PCS calculation but save on a couple of potential scaling roundtrips.

Also, I'm aware that in general I'm flopping between e.g. `Vector` and `CieXyz` quite a bit, is that problematic?

Reply:
We should try to avoid switching between the types where possible as that causes per-pixel overhead. Perhaps we should be trying to provide this information up-front? It's tricky...

Reply:
I'm a little confused by this comment?

```c#
// when data to PCS, output from calculator is descaled XYZ
// but expected return value is scaled XYZ
// see DemoMaxICC IccCmm.cpp : CIccXformMatrixTRC::Apply()
```

How does the calculator know how to descale the output from the scaled input?

Reply:
I'm not sure I follow your question but I'm still getting my head around it myself. Does it help if I say...
- Matrix TRC can only be used with XYZ PCS
- Simple curve lookups use the scaled XYZ
- In this case, because the output of the curve needs to be multiplied with a matrix, we need to use the actual descaled XYZ it represents
- Downstream operations tend to assume the PCS will be in scaled form

I've experimented with a different approach that might make things clearer, moving the responsibility of knowing about this from `ColorTrcCalculator` to `ColorProfileConverter`. The downside is you can't "just use" the calculator, but that's true of ICC transforms generally anyway once you start looking closely.

Imagine `ColorTrcCalculator` doesn't have these changes, and instead `ColorProfileConverter` has these extra steps:

```c#
// output of Matrix TRC calculator is descaled XYZ, needs to be re-scaled to be used as PCS
Vector4 sourcePcs = sourceParams.Converter.Calculate(source.ToScaledVector4());
sourcePcs = sourceParams.IsMatrixTrc ? new CieXyz(sourcePcs.AsVector3()).ToScaledVector4() : sourcePcs;
```

```c#
Vector4 targetPcs = ... // do conversion from source PCS to target PCS, potentially with perceptual adjustments

// input to Matrix TRC calculator is descaled XYZ, need to descale PCS before use
targetPcs = targetParams.IsMatrixTrc ? new Vector4(CieXyz.FromScaledVector4(targetPcs).ToVector3(), 1) : targetPcs;
return TTo.FromScaledVector4(targetParams.Converter.Calculate(targetPcs));
```

I can't tell if it feels better or worse. Depends what mood I'm in when I look!

As for Gray TRC, I can't see any evidence that it does the same thing, just multiplies the scaled monochrome by scaled D50.

---

💡 If [that matrix multiplication](https://github.com/SixLabors/ImageSharp/pull/1567#discussion_r1889208728) in `LutEntryCalculator` does indeed need fixing, then presumably the matrix multiplication in `ColorTrcCalculator` does too.